File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    public List<String> l;
    public String s;
    public int i;
    public HashMap<Character, String> hm;
    
    public List<String> letterCombinations(String digits) {
        l = new ArrayList<>();
        s = "";
        i = 0;
        hm = new HashMap<>();
        hm.put('2', "abc");
        hm.put('3', "def");
        hm.put('4', "ghi");
        hm.put('5', "jkl");
        hm.put('6', "mno");
        hm.put('7', "pqrs");
        hm.put('8', "tuv");
        hm.put('9', "wxyz");
        if (digits.length() == 0)
        {
            return l;
        }
        
        doCombinations(digits);
        return l;
    }
    
    public void doCombinations(String digits)
    {
        if (s.length() == digits.length())
        {
            l.add(s);
            return;
        }
        String a = hm.get(digits.charAt(i));
        for (int j = 0; j<a.length();j++)
        {
            s = s + a.charAt(j);
            i++;
            doCombinations(digits);
            
            s = s.substring(0, s.length() - 1);
            i--;
        }
    }
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if len(digits) == 0:
            return []
        res = []
        string = ""
        num = {
            2: ["a", "b", "c"],
            3: ["d", "e", "f"],
            4: ["g", "h", "i"],
            5: ["j", "k", "l"],
            6: ["m", "n", "o"],
            7: ["p", "q", "r", "s"],
            8: ["t", "u", "v"],
            9: ["w", "x", "y", "z"]
        }
        def dp(i):
            nonlocal string, res, num
            if len(string) == len(digits):
                temp = string
                res.append(temp)
                return
            for j in range(len(num[int(digits[i])])):
                string += num[int(digits[i])][j]
                dp(i + 1)
                string = string[:-1]
            return
        dp(0)
        return res